test/es-module/test-esm-dynamic-import.js

```diff
@@ -0,0 +1,96 @@
+// Flags: --experimental-modules --harmony-dynamic-import
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+const { URL } = require('url');
+const vm = require('vm');
+
+common.crashOnUnhandledRejection();
+
+const relativePath = './test-esm-ok.mjs';
+const absolutePath = require.resolve('./test-esm-ok.mjs');
+const targetURL = new URL('file:///');
+targetURL.pathname = absolutePath;
+
+function expectErrorProperty(result, propertyKey, value) {
+  Promise.resolve(result)
+    .catch(common.mustCall(error => {
+      assert.equal(error[propertyKey], value);
+    }));
+}
+
+function expectMissingModuleError(result) {
+  expectErrorProperty(result, 'code', 'ERR_MISSING_MODULE');
+}
+
+function expectInvalidUrlError(result) {
+  expectErrorProperty(result, 'code', 'ERR_INVALID_URL');
+}
+
+function expectInvalidProtocolError(result) {
+  expectErrorProperty(result, 'code', 'ERR_INVALID_PROTOCOL');
+}
+
+function expectInvalidContextError(result) {
+  expectErrorProperty(result, 'message', 'invalid_context');
+}
+
+function expectOkNamespace(result) {
+  Promise.resolve(result)
+    .then(common.mustCall(ns => {
+      // Can't deepStrictEqual because ns isn't a normal object
+      assert.deepEqual(ns, { default: true });
+    }));
+}
+
+function expectFsNamespace(result) {
+  Promise.resolve(result)
+    .then(common.mustCall(ns => {
+      assert.equal(typeof ns.default.writeFile, 'function');
+    }));
+}
+
+expectOkNamespace(import(relativePath));
+
+expectFsNamespace(import("fs"));
+
+expectInvalidProtocolError(import("node:fs"));
+
+// Fails, no valid referrer (indirect eval) & relative path
+expectMissingModuleError(eval.call(null, `import("${relativePath}")`));
+
+// Fails, no valid referrer (direct eval) & relative path
+expectMissingModuleError(eval(`import("${relativePath}")`));
+
+// Succeeds, doesn't need a valid referrer
+expectFsNamespace(eval.call(null, 'import("fs")'));
+
+// Fails because referrer is undefined (and loader/search.js rejects it)
+// TODO(jkrems): This doesn't seem like expected behavior
+expectMissingModuleError(eval.call(null, `import("${targetURL}")`));
+
+// Succeeds
+expectFsNamespace(vm.runInThisContext('import("fs")'));
+
+// Succeeds because it's got an valid base url
+expectOkNamespace(vm.runInThisContext(`import("${relativePath}")`, {
+  filename: __filename,
+}));
+
+// Rejects because it's got an invalid referrer URL
+expectInvalidUrlError(vm.runInThisContext(`import("${relativePath}")`));
+
+// Rejects because it's got an invalid referrer URL
+expectInvalidUrlError(vm.runInThisContext(`import("${targetURL}")`));
+
+// Rejects because it's running in the wrong context
+expectInvalidContextError(
```

Comment:
Double-checked in Chrome and at least for `<iframe>`, each context gets it's own module map. Which I guess would be possible in the future for node as well, but for now throwing seems fair - at least as the default behavior.